### PR TITLE
Adding ServerConfiguration

### PR DESCRIPTION
**Pull request by joesus**
## Checklist

- [x] I've read the [Contributing Guidelines](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md)
- [x] I've completed the [Contributor License Agreement](https://developers.facebook.com/opensource/cla)
- [x] I've ensured that all existing tests pass and added tests (when/where necessary)
- [x] I've ensured that my code lints properly: (`swiftlint && swiftlint autocorrect --format`)
- [x] I've updated the documentation (when/where necessary) and [Changelog](CHANGELOG.md) (when/where necessary)
- [x] I've added the proper label to this pull request (e.g. `bug` for bug fixes)

## Pull Request Details

Introduces ServerConfiguration. 
Introduces related types for `DialogFlow`, `DialogConfiguration`, and `RestrictiveRule`
This is still missing the implementation for `EventBinding` as well as a service for providing these objects.

See commit history for more detail.

## Test Plan

Unit tests should cover everything added.

### CODE REVIEW

File: Sources/Core/Extensions/String+Extensions.swift [hunk 1]
```diff
@@ -79,4 +79,15 @@ extension String {
 
     return MathUtility.ceil(for: size)
   }
+
+  /**
+   Returns nil is the String is empty. Returns the optional String only if it is non-empty.
+   */
+  var nonEmpty: String? {
```
Comment (codytwinton) [inline Sources/Core/Extensions/String+Extensions.swift]: non empty kinda implies boolean result. [Nonempty](https://en.wiktionary.org/wiki/nonempty)

```suggestion
  var nonempty: String? {
```

File: Sources/Core/RemoteRestrictiveEventParameter.swift [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+// swiftlint:disable discouraged_optional_boolean discouraged_optional_collection
+
+import Foundation
+
+struct RemoteRestrictiveEventParameter: Decodable {
```
Comment (codytwinton) [inline Sources/Core/RemoteRestrictiveEventParameter.swift]: Change to

```swift
extension Remote {
  struct RestrictiveEventParameter: Decodable {

File: Sources/Core/Extensions/String+Extensions.swift (hunk 1 above)
Comment (joesus) [reply Sources/Core/Extensions/String+Extensions.swift]: Yeah... I didn't love this name... do you have ideas on what would make it more clear? Really it's more like a `validatedForNonEmptiness` but that seemed too verbose... probably a dependent type would make more sense. Thoughts?

File: Sources/Core/RemoteRestrictiveEventParameter.swift (hunk 2 above)
Comment (joesus) [reply Sources/Core/RemoteRestrictiveEventParameter.swift]: Oh. Yes. I merged the branch with the remote namespacing into the follow up pr to this one and have these fixed in there.

**Pull request merged**